File: physics/validator.py
"""Top-level physics validation entry point: `validate_layout(scene) -> dict`.

Integrates every sibling module into one call: `containment.check_scene_containment`,
`collision.check_collision` (all object pairs, pruned by `collision.aabb_overlap`
broad-phase), `support.check_support`, and `constraints.check_constraints`. This is
the only function the packing-solver / renderer teammates should call directly.

Malformed-input contract
-------------------------
This function never raises. Before running any check it validates scene geometry
up front (`check_no_duplicate_ids` + `obb_from` on the container and every object).
If that raises `ValueError` (NaN/non-finite dims or position, bad quaternion,
duplicate ids), it is caught and a `MALFORMED_GEOMETRY` violation is returned
immediately with `score=0.0`, `valid=False` -- the rest of the pipeline never
runs on data it can't trust.

Output shape
------------
    {
      "valid": bool,        # True iff `violations` is empty
      "score": float,       # 1.0 clean -> lower as violations pile up/worsen
      "violations": [...],  # hard failures, see below
      "warnings": [...],    # soft/informational, do not affect `valid`/`score`
    }

Violation types: OBJECT_COLLISION, CONTAINER_PENETRATION, UNSUPPORTED_OBJECT,
FRAGILE_OBJECT_OVERLOADED, LIQUID_NOT_UPRIGHT, INVALID_ORIENTATION,
MALFORMED_GEOMETRY. Warning types: UNSTABLE_STACK, FRAGILE_LOAD, HEAVY_ON_TOP,
SOFT_COMPRESSION (see compressibility policy below).

Hard-fail vs warning policy
----------------------------
- A completely unsupported object (`SupportResult.floating`, i.e.
  `support_ratio < floating_threshold`) is a hard-fail `UNSUPPORTED_OBJECT`
  violation -- an item with (near-)zero contact area under gravity is not a
  packable layout.
- An object that IS supported (`support_ratio >= floating_threshold`) but whose
  center of mass falls outside its support footprint (`SupportResult.unstable`)
  is only a `UNSTABLE_STACK` warning -- it is precariously balanced, not
  physically impossible, and `valid` can still be True for such a scene. This
  is a deliberate simplification (no friction/tip-over dynamics modeled here,
  see support.py) -- treat repeated UNSTABLE_STACK warnings as "worth a second
  look", not "reject".
- Collisions and container-wall penetrations are always hard fails (the
  geometry is physically impossible) -- EXCEPT that raw overlap/penetration is
  first reduced by a compressibility allowance (`physics.compressibility`,
  derived from each object's `rigidity`/`compressibility_k`) before that
  decision is made: if the allowance fully absorbs the raw overlap the pair
  becomes a `SOFT_COMPRESSION` warning instead of a violation (rigid objects
  get zero allowance, so this never changes behavior for rigid-only scenes);
  if some overlap remains past what compression can plausibly absorb, it's
  still a hard-fail OBJECT_COLLISION/CONTAINER_PENETRATION, but severity and
  `penetration_depth_m` use only that excess, not the raw depth.
  LIQUID_NOT_UPRIGHT / INVALID_ORIENTATION / FRAGILE_OBJECT_OVERLOADED
  passthrough from `check_constraints` stay hard fails (they are already
  violations there); FRAGILE_LOAD / HEAVY_ON_TOP stay warnings (already
  warnings there).

Severity heuristics (each in [0, 1], documented per type; used only for score)
--------------------------------------------------------------------------------
- OBJECT_COLLISION: penetration_depth_m / min(smallest full extent of A, smallest
  full extent of B) -- a penetration comparable to an object's own thinnest
  dimension is maximally severe.
- CONTAINER_PENETRATION: penetration_depth_m / smallest container half-extent.
- UNSUPPORTED_OBJECT: 1 - support_ratio / floating_threshold -- 1.0 for a fully
  floating object (support_ratio=0), tapering to 0 as support_ratio approaches
  the floating threshold from below.
- FRAGILE_OBJECT_OVERLOADED: supported_weight_kg / max(0.1, object's own mass_kg)
  -- weight piled on a fragile item relative to its own mass.
- LIQUID_NOT_UPRIGHT / INVALID_ORIENTATION: (tilt_deg - tolerance_deg) / (90 -
  tolerance_deg) -- 0 right at the tolerance boundary, 1.0 at a full 90 degree
  tilt (the worst case for these angle metrics).
- MALFORMED_GEOMETRY: no severity (short-circuits with score=0.0 regardless).

Score formula
--------------
`score = max(0, 1 - sum(min(1, severity) for v in violations) / max(1, n))`
where `n = len(scene.objects)`. Warnings never affect score. Monotonic:
adding a violation or increasing any severity can only lower or hold the score.

Determinism
------------
No randomness. `violations` and `warnings` are each sorted by `(type, object_id)`
(the pair's lexicographically-smaller id for OBJECT_COLLISION) before being
returned, so `json.dumps(result, sort_keys=False)` is byte-identical across
repeated calls on the same scene.

Complexity
-----------
O(n^2) overall, dominated by the all-pairs collision scan (n = object count):
each pair is pruned first by O(1) `aabb_overlap`, and only surviving pairs run
full O(1) SAT via `check_collision` -- so the O(n^2) factor is cheap per-pair
broad-phase work, not O(n^2) SAT. `check_scene_containment` is O(n),
`check_support` and `check_constraints` are each O(n^2) internally (same scale
as the collision scan, see their own docstrings) -- fine at hackathon scale
(5-20 objects); a spatial index would be the upgrade path beyond that.
"""
from __future__ import annotations

import numpy as np

from physics.collision import aabb_overlap, check_collision
from physics.compressibility import (
    axis_projected_extent_m,
    combined_collision_allowance_m,
    container_wall_allowance_m,
)
from physics.constraints import check_constraints
from physics.containment import check_no_duplicate_ids, check_scene_containment
from physics.geometry import obb_from
from physics.schema import Scene
from physics.support import check_support

# wall name suffix ("+x" / "-x" -> "x") -> container-local axis index
_WALL_AXIS_INDEX = {"x": 0, "y": 1, "z": 2}

DEFAULT_FLOATING_THRESHOLD = 0.05


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, float(x)))


def _vec3(v) -> list[float]:
    return [float(v[0]), float(v[1]), float(v[2])]


def _sort_key(entry: dict):
    oid = entry.get("object")
    if oid is None:
        objs = entry.get("objects")
        oid = min(objs) if objs else ""
    return (entry["type"], oid)


def _malformed(object_id, exc: ValueError) -> dict:
    return {
        "valid": False,
        "score": 0.0,
        "violations": [
            {"type": "MALFORMED_GEOMETRY", "object": object_id, "detail": str(exc)}
        ],
        "warnings": [],
    }


def validate_layout(scene: Scene) -> dict:
    """Run the full physics validation pipeline on `scene`. See module docstring."""
    # Step 1: malformed-input guard -- never let a bad scene raise past here.
    try:
        check_no_duplicate_ids(scene)
    except ValueError as e:
        return _malformed(None, e)

    try:
        container_obb = obb_from(scene.container)
    except ValueError as e:
        return _malformed(scene.container.id, e)

    obbs = {}
    try:
        for obj in scene.objects:
            obbs[obj.id] = obb_from(obj)
    except ValueError as e:
        return _malformed(obj.id, e)

    n = max(1, len(scene.objects))
    violations: list[dict] = []
    warnings: list[dict] = []

    # --- Containment ---
    objs = scene.objects
    obj_by_id = {o.id: o for o in objs}
    container_denom = max(1e-9, float(np.min(container_obb.half_extents)))
    for res in check_scene_containment(scene):
        raw_depth = res.penetration_depth_m
        allowance = 0.0
        if res.violated_walls:
            # check_containment only reports one scalar penetration_depth_m
            # (already the max across every violated vertex/axis), not a
            # per-wall breakdown, so there's no way to pick "the wall with
            # the largest raw penetration" from ContainmentResult alone
            # without touching containment.py (out of scope for this task).
            # Resolution: use the axis of the first violated wall in sorted
            # (deterministic) order as the stand-in axis for the object's
            # own compressible extent -- simple, deterministic, and the
            # `raw_depth` used below is still the true overall max.
            axis_idx = _WALL_AXIS_INDEX[sorted(res.violated_walls)[0][-1]]
            axis = container_obb.axes[:, axis_idx]
            obj = obj_by_id[res.object_id]
            extent = axis_projected_extent_m(obbs[res.object_id], axis)
            allowance = container_wall_allowance_m(obj, extent)
        effective_depth = max(0.0, raw_depth - allowance)
        if effective_depth <= 0.0:
            if raw_depth > 0.0:
                warnings.append(
                    {
                        "type": "SOFT_COMPRESSION",
                        "object": res.object_id,
                        "raw_penetration_depth_m": raw_depth,
                        "compressed_depth_m": raw_depth,
                    }
                )
            continue
        severity = _clamp01(effective_depth / container_denom)
        violations.append(
            {
                "type": "CONTAINER_PENETRATION",
                "object": res.object_id,
                "penetration_depth_m": effective_depth,
                "violated_walls": list(res.violated_walls),
                "severity": severity,
            }
        )

    # --- Collisions: all pairs, aabb_overlap broad-phase first (O(n^2) total). ---
    for i in range(len(objs)):
        for j in range(i + 1, len(objs)):
            a, b = objs[i], objs[j]
            oa, ob = obbs[a.id], obbs[b.id]
            if not aabb_overlap(oa, ob):
                continue
            result = check_collision(oa, ob)
            if result.colliding:
                raw_depth = result.penetration_depth_m
                allowance = combined_collision_allowance_m(a, b, result.axis, oa, ob)
                effective_depth = max(0.0, raw_depth - allowance)
                if effective_depth <= 0.0:
                    if raw_depth > 0.0:
                        warnings.append(
                            {
                                "type": "SOFT_COMPRESSION",
                                "objects": sorted([a.id, b.id]),
                                "raw_penetration_depth_m": raw_depth,
                                "compressed_depth_m": raw_depth,
                            }
                        )
                    continue
                denom = max(1e-9, min(min(a.dimensions), min(b.dimensions)))
                severity = _clamp01(effective_depth / denom)
                violations.append(
                    {
                        "type": "OBJECT_COLLISION",
                        "objects": sorted([a.id, b.id]),
                        "penetration_depth_m": effective_depth,
                        "contact_point": _vec3(result.contact_point),
                        "severity": severity,
                    }
                )

    # --- Support ---
    for r in check_support(scene, floating_threshold=DEFAULT_FLOATING_THRESHOLD):
        if r.floating:
            severity = _clamp01(1.0 - r.support_ratio / DEFAULT_FLOATING_THRESHOLD)
            violations.append(
                {
                    "type": "UNSUPPORTED_OBJECT",
                    "object": r.object_id,
                    "support_ratio": r.support_ratio,
                    "severity": severity,
                }
            )
        elif r.unstable:
            center = obbs[r.object_id].center
            warnings.append(
                {
                    "type": "UNSTABLE_STACK",
                    "object": r.object_id,
                    "stability_margin_m": r.stability_margin_m,
                    "center_of_mass_projection": [float(center[0]), float(center[2])],
                }
            )

    # --- Constraints (travel semantics) ---
    mass_by_id = {o.id: o.mass_kg for o in objs}
    c_violations, c_warnings = check_constraints(scene)
    for v in c_violations:
        if v.type == "FRAGILE_OBJECT_OVERLOADED":
            weight = v.details.get("supported_weight_kg", 0.0)
            denom = max(0.1, mass_by_id.get(v.object_id, 0.1))
            severity = _clamp01(weight / denom)
            violations.append(
                {
                    "type": v.type,
                    "object": v.object_id,
                    "supported_weight_kg": weight,
                    "severity": severity,
                }
            )
        else:  # LIQUID_NOT_UPRIGHT / INVALID_ORIENTATION
            tilt = v.details.get("tilt_deg", 0.0)
            tol = v.details.get("tolerance_deg", 15.0)
            denom = max(1e-9, 90.0 - tol)
            severity = _clamp01((tilt - tol) / denom)
            entry = {"type": v.type, "object": v.object_id, "severity": severity}
            entry.update(v.details)
            violations.append(entry)

    for w in c_warnings:
        entry = {"type": w.type, "object": w.object_id}
        entry.update(w.details)
        warnings.append(entry)

    violations.sort(key=_sort_key)
    warnings.sort(key=_sort_key)

    severity_sum = sum(min(1.0, v.get("severity", 1.0)) for v in violations)
    score = max(0.0, 1.0 - severity_sum / n)

    return {
        "valid": len(violations) == 0,
        "score": score,
        "violations": violations,
        "warnings": warnings,
    }

File: swift/PackPhysics/Sources/PackPhysics/Validator.swift
// Top-level physics validation entry point. Port of physics/validator.py.
//
// One call runs the whole deterministic pipeline on a candidate layout:
//
//     precompute (SceneGeometry.swift)     one pass: OBBs, vertices, AABBs
//       -> containment  (per-wall depths, per-wall compressibility)
//       -> collision    (AABB broad phase + 15-axis SAT)
//       -> support      (exact contact polygons, static stability, chains)
//       -> constraints  (travel metadata, transitive load)
//       -> metrics      (COM, fill, clearances -- solver objective terms)
//
// This is the only function the packing solver / renderer needs to call. For a
// solver inner loop ("can I add this one object?") use `PlacementValidator`
// (Incremental.swift) and call this once at the end.
//
// Never throws. Any `MalformedSceneError` (non-finite / non-positive dims,
// non-finite position, zero quaternion, duplicate ids) short-circuits to
// `{valid: false, score: 0, violations: [MALFORMED_GEOMETRY], warnings: [],
// metrics: {}}`.
//
// Determinism: no randomness; `violations` and `warnings` are each sorted by
// `(type, objectKey)` where a pair entry uses its lexicographically smaller id.
// Python's `list.sort` is stable, Swift's `sort` is not, so the original
// insertion index is the final tie-breaker here -- byte-identical ordering.
//
// Severity heuristics (each clamped to [0,1]; used only for `score`) and the
// exact per-entry field tables are documented in docs/PHYSICS.md §10.

import Foundation

/// Wall name suffix ("+x" / "-x" -> "x") -> container-local axis index.
private let wallAxisIndex: [Character: Int] = ["x": 0, "y": 1, "z": 2]

public let defaultFloatingThreshold = 0.05

@inline(__always) private func clamp01(_ x: Double) -> Double { max(0.0, min(1.0, x)) }

@inline(__always) private func minComponent(_ v: Vec3) -> Double { min(v.x, min(v.y, v.z)) }

/// The MALFORMED_GEOMETRY short-circuit result (`metrics` is an empty object,
/// matching Python v2's `"metrics": {}`).
private func malformedResult(_ objectId: String?, _ detail: String) -> ValidationResult {
    ValidationResult(
        valid: false,
        score: 0.0,
        violations: [.object([
            "type": "MALFORMED_GEOMETRY",
            "object": objectId.map { JSONValue.string($0) } ?? .null,
            "detail": .string(detail),
        ])],
        warnings: [],
        metrics: .object([:])
    )
}

/// Python `_sort_key`: `(type, object)`, or for a pair entry the
/// lexicographically smaller of `objects`.
private func sortKey(_ entry: JSONValue) -> (String, String) {
    let type = entry["type"]?.stringValue ?? ""
    if let oid = entry["object"]?.stringValue { return (type, oid) }
    if let objs = entry["objects"]?.arrayValue, !objs.isEmpty {
        return (type, objs.compactMap { $0.stringValue }.min() ?? "")
    }
    return (type, "")
}

/// Stable sort by `sortKey` -- equal keys keep insertion order, like Python.
private func sortedEntries(_ entries: [JSONValue]) -> [JSONValue] {
    entries.enumerated().sorted { l, r in
        let (a, b) = (sortKey(l.element), sortKey(r.element))
        return a == b ? l.offset < r.offset : a < b
    }.map(\.element)
}

/// Run the full physics validation pipeline on `scene`. Never throws.
public func validateLayout(_ scene: Scene,
                           floatingThreshold: Double = defaultFloatingThreshold) -> ValidationResult {
    do {
        return try pipeline(scene, floatingThreshold: floatingThreshold)
    } catch let e as MalformedSceneError {
        return malformedResult(e.objectId, e.message)
    } catch {
        // Defensive: anything geometry raises that isn't tagged with an id.
        return malformedResult(nil, "\(error)")
    }
}

private func pipeline(_ scene: Scene, floatingThreshold: Double) throws -> ValidationResult {
    let geom = try precompute(scene)
    let objs = geom.objects
    let n = max(1, geom.n)
    var violations: [JSONValue] = []
    var warnings: [JSONValue] = []
    let container = geom.containerOBB

    // --- Containment (per-wall compressibility) ---
    let containerDenom = max(1e-9, minComponent(container.halfExtents))
    for res in try checkSceneContainment(scene, geom: geom) {
        let i = geom.index[res.objectId]!
        let obj = objs[i]
        let rawDepth = res.penetrationDepthM
        var effective: [String: Double] = [:]
        if obj.rigidity == .rigid {
            effective = res.perWallDepthM  // rigid -> allowance 0
        } else {
            for (wall, depth) in res.perWallDepthM {
                let axis = container.axes.column(wallAxisIndex[wall.last!]!)
                let allowance = containerWallAllowanceM(
                    obj, wallAxisExtentM: axisProjectedExtentM(geom.obbs[i], axis: axis))
                effective[wall] = max(0.0, depth - allowance)
            }
        }
        let effectiveDepth = effective.values.max() ?? 0.0
        if effectiveDepth <= 0.0 {
            if rawDepth > 0.0 {
                warnings.append(.object([
                    "type": "SOFT_COMPRESSION",
                    "object": .string(res.objectId),
                    "raw_penetration_depth_m": .number(rawDepth),
                    "compressed_depth_m": .number(rawDepth),
                    "walls": .strings(res.perWallDepthM.keys.sorted()),
                ]))
            }
            continue
        }
        var perWall: [String: JSONValue] = [:]
        for (w, d) in effective where d > 0.0 { perWall[w] = .number(d) }
        violations.append(.object([
            "type": "CONTAINER_PENETRATION",
            "object": .string(res.objectId),
            "penetration_depth_m": .number(effectiveDepth),
            "raw_penetration_depth_m": .number(rawDepth),
            "compressed_depth_m": .number(rawDepth - effectiveDepth),
            "violated_walls": .strings(perWall.keys.sorted()),
            "per_wall_depth_m": .object(perWall),
            "penetrating_vertices": .array(res.penetratingVertices.map { .vec($0) }),
            "severity": .number(clamp01(effectiveDepth / containerDenom)),
        ]))
    }

    // --- Collisions (broad phase + SAT; only colliding pairs come back) ---
    for r in collideScene(geom) {
        let i = geom.index[r.aId]!, j = geom.index[r.bId]!
        let a = objs[i], b = objs[j]
        let rawDepth = r.penetrationDepthM
        let mtvAxis = r.axis ?? Vec3(0, 0, 0)  // always set when colliding
        let contactPoint = r.contactPoint ?? Vec3(0, 0, 0)
        let allowance: Double
        if a.rigidity == .rigid && b.rigidity == .rigid {
            allowance = 0.0
        } else {
            allowance = combinedCollisionAllowanceM(
                a, b, mtvAxis: mtvAxis, obbA: geom.obbs[i], obbB: geom.obbs[j])
        }
        let effectiveDepth = max(0.0, rawDepth - allowance)
        let pair = [a.id, b.id].sorted()
        if effectiveDepth <= 0.0 {
            if rawDepth > 0.0 {
                warnings.append(.object([
                    "type": "SOFT_COMPRESSION",
                    "objects": .strings(pair),
                    "raw_penetration_depth_m": .number(rawDepth),
                    "compressed_depth_m": .number(rawDepth),
                    "contact_point": .vec(contactPoint),
                ]))
            }
            continue
        }
        let denom = max(1e-9, min(minComponent(a.dimensions), minComponent(b.dimensions)))
        violations.append(.object([
            "type": "OBJECT_COLLISION",
            "objects": .strings(pair),
            "penetration_depth_m": .number(effectiveDepth),
            "raw_penetration_depth_m": .number(rawDepth),
            "compressed_depth_m": .number(rawDepth - effectiveDepth),
            "contact_point": .vec(contactPoint),
            "axis": .vec(mtvAxis),
            "severity": .number(clamp01(effectiveDepth / denom)),
        ]))
    }

    // --- Support / static stability ---
    for s in try checkSupport(scene, floatingThreshold: floatingThreshold, geom: geom) {
        let center = geom.obbs[geom.index[s.objectId]!].center
        let comXZ = JSONValue.array([.number(center.x), .number(center.z)])
        if s.floating {
            violations.append(.object([
                "type": "UNSUPPORTED_OBJECT",
                "object": .string(s.objectId),
                "support_ratio": .number(s.supportRatio),
                "center_of_mass_projection": comXZ,
                "severity": .number(clamp01(1.0 - s.supportRatio / floatingThreshold)),
            ]))
        } else if s.unstable {
            warnings.append(.object([
                "type": "UNSTABLE_STACK",
                "object": .string(s.objectId),
                "stability_margin_m": .number(s.stabilityMarginM),
                "support_ratio": .number(s.supportRatio),
                "supporting_objects": .strings(s.supportingObjects),
                "center_of_mass_projection": comXZ,
                "contact_polygon": .array(s.contactPolygon.map { p in .array(p.map { .number($0) }) }),
            ]))
        } else if s.supportedByUnstable {
            warnings.append(.object([
                "type": "UNSTABLE_SUPPORT_CHAIN",
                "object": .string(s.objectId),
                "supporting_objects": .strings(s.supportingObjects),
            ]))
        }
    }

    // --- Travel constraints (transitive load) ---
    var massById: [String: Double] = [:]
    for o in objs { massById[o.id] = o.massKg }
    let (cViolations, cWarnings) = try checkConstraints(scene, geom: geom)
    for v in cViolations {
        var entry: [String: JSONValue] = ["type": .string(v.type), "object": .string(v.objectId)]
        for (k, val) in v.details { entry[k] = val }
        if v.type == "FRAGILE_OBJECT_OVERLOADED" {
            let weight = v.details["supported_weight_kg"]?.doubleValue ?? 0.0
            entry["severity"] = .number(clamp01(weight / max(0.1, massById[v.objectId] ?? 0.1)))
        } else {  // LIQUID_NOT_UPRIGHT / INVALID_ORIENTATION
            let tilt = v.details["tilt_deg"]?.doubleValue ?? 0.0
            let tol = v.details["tolerance_deg"]?.doubleValue ?? 15.0
            entry["severity"] = .number(clamp01((tilt - tol) / max(1e-9, 90.0 - tol)))
        }
        violations.append(.object(entry))
    }
    for w in cWarnings {
        var entry: [String: JSONValue] = ["type": .string(w.type), "object": .string(w.objectId)]
        for (k, val) in w.details { entry[k] = val }
        warnings.append(.object(entry))
    }

    violations = sortedEntries(violations)
    warnings = sortedEntries(warnings)
    // Summed in sorted order, like Python (float addition is not associative).
    var severitySum = 0.0
    for v in violations { severitySum += min(1.0, v["severity"]?.doubleValue ?? 1.0) }

    return ValidationResult(
        valid: violations.isEmpty,
        score: max(0.0, 1.0 - severitySum / Double(n)),
        violations: violations,
        warnings: warnings,
        metrics: sceneMetrics(geom)
    )
}

/// `validate(scene, placements)`: apply `placements` (if given), then
/// `validateLayout`. Never throws -- an unknown placement id comes back as the
/// single MALFORMED_GEOMETRY violation naming the first bad id, instead of
/// propagating `applyPlacements`'s error. Mirrors `physics.io.validate`.
public func validate(_ scene: Scene, placements: [Placement]? = nil) -> ValidationResult {
    guard let placements = placements else { return validateLayout(scene) }
    do {
        return validateLayout(try applyPlacements(scene, placements))
    } catch let e as MalformedSceneError {
        return malformedResult(e.objectId, e.message)
    } catch {
        return malformedResult(nil, "\(error)")
    }
}
